### ISSUE TITLE
Creating a way to pass objects from Server to Document

### ISSUE THREAD

**Issue by adam-hubbell** [bokeh/bokeh#5138]
I was considering adding to bokeh a way to pass data from a server(one started programmatically) to any bokeh applications running on that server. We already implement all of this functionality on our server, but from outside bokeh in a roundabout way that involves some monkeypatching. 
How it would work:
-user starts a server like so:
```
from bokeh.server.server import Server
from bokeh.command.util import build_single_handler_applications
#app_path_dict is a dictionary with paths to bokeh files/directories as the key, and None as the value
apps = build_single_handler_applications(app_path_dict.keys(),app_path_dict)
srv = Server(apps, some_kwargs)
```
-Then they would send objects to a document running on the server by calling something like:
```
srv.send_object_to_server('my_documents_name',the_object_im_passing,'a_key_for_the_object')
#my_documents_name is the same string that would be used to call that document from 
#autoload_server's app_path paramater. (maybe without the leading slash?)
```
-In the document code itself, you would access that object by calling something along the lines of:
```
my_object_from_server = curdoc().get_object_from_server('a_key_for_the_object')
```
-then for receiving the object in an already active session, the user define their own function for handling the new object like so:
```
@coroutine
def live_data_func():
     my_column_data_source.data = curdoc().get_object_from_server('a_key_for_the_object')
curdoc().add_live_data_callback(live_data_func) #gets called on insert data
```
To implement this, I would need to add the previously described functions to Document and Server, I would need to slightly alter session_context to pass it's instance of server_context to it's _document object, and alter Document to receive this this server_context, so that Server and Document have a common object that they can both see. 
Is this something you would like to see implemented? Any input on method names, implementation, or anything else would also be much appreciated. Thank you.

**Comment by bryevdv** [comment]
@adam-hubbell so far this is fairly abstract, for my benefit, can you describe a concrete use-case?

**Comment by gryBox** [comment]
Here are some of the uses

- passing server configs. e.g. ports, ips ...
- starting and shutting down the server programmatically
- passing data(?)

@adam-hubbell  Is there anything else we use it for?  
I was told by @birdsarah  that @mrocklin uses a similar method (with `bokeh`) for computation visuals in `distributed` and `dask`.

**Comment by mrocklin** [comment]
For breadth, the strategy used in dask.distributed is to maintain global state in a module that is accessible by both the server_lifecycle.py and main.py scripts.  We set up a coroutine in server_lifecycle that continuously gets data from an external source and updates the global state.  In main.py we set up a bokeh-periodic-callback to query that data and update the columndatasource.  

What I would *like* to do is to use normal tornado callbacks within main.py so that I could, for example, use Tornado Events to trigger actions in all of the main.py instances rather than use Bokeh's PeriodicCallbacks.  This would increase responsiveness and probably decrease extraneous load.  This currently doesn't work because Bokeh uses locks and so doesn't play nicely with normal tornado coroutines.

As someone familiar with Tornado I'm generally of the opinion that Bokeh should not add any more special callback mechanisms, but should instead get out of the way so that existing Tornado machinery can work.  I don't think that the Bokeh team should get into the business of maintaining a wealth of special concurrency mechanisms.

We can remove the locks in the Bokeh server as long as we always present valid state between any two yield points.  When I was looking at this problem a while ago it seemed doable with some effort.

I do think that something like what @gryBox suggests might be nice.  It could be nice to have a single obvious place to put global state and events to wait on for updates.  Personally I would probably wait until there were a few people managing global state before fixing what Bokeh's maintained strategy would be.

**Comment by adam-hubbell** [comment]
One of the main reasons we use it is to avoid certain redundancies when we have multiple sessions open. For example, we have a few csvs that we poll, watching to see if the modify time has changed. Then when it has, we read the data into a pandas dataframe, do some calculations with said dataframe, then pass it to a bokeh document to be put into a data table. If we're doing this with periodic callbacks, or a separate thread in the document with a next_tick_callback, we'll do the same reading and calculating for every session that's open. But, using the above strategy, we would have the reading and calculating happen only once, and then passed into the already running sessions. It also avoids the redundancy of passing the same data to multiple different documents running on the same server.

**Comment by bigreddot** [comment]
ping @havocp

**Comment by havocp** [comment]
i don't have a developed suggestion here but it strikes me as strained that the Document is involved in essentially using a raw websocket, in the original proposal here. I'd probably put an API like that on Session instead (and make it more or less directly map to websockets perhaps). The Document's concern should strictly be that it's a tree of models, surely, don't want the kitchen sink in that class.

cleaning up locking seems like a distinct problem but maybe I'm confused on that. I don't remember the discussion we had about that previously very well but hopefully its saved on another issue somewhere :-)

**Comment by adam-hubbell** [comment]
There would be no web socket usage in the implementation, it would use preexisting bokeh functionality to accomplish. I see that document has an instance of session_context that's set by ApplicationContext, and I could use that. The code in Document itself would be very minimal, but I figured calling curdoc().some_method_to_get_object() would make it easier on the user. I'm not yet sure where the object would live until the document wants it, maybe application_context or server_context. The whole process of passing an object would be:
instance_of_Server.pass_object('app_name','key',object) #in the server code
-----------------------------------------
curdoc().get_object('key') #in the document code

Where pass_object sends the object to go live in a dict somewhere, and get_object would look in that dict.

**Comment by havocp** [comment]
Oh I misunderstood what you meant. Still I wouldnt get document involved - maybe add something like curserver() - I think I used to have an idea about that but I forgot it right now.

also there's an email thread I've linked from a couple other issues where I suggested an API where youd essentially provide a couple functions, one to fetch data from somehwere and another to be called with each session and the fetched data as args. And then bokeh would do the machinery around those two, to fetch data and update all sessions with it.

**Comment by gryBox** [comment]
>  I suggested an API where youd essentially provide a couple functions, one to fetch data from somehwere and another to be called with each session and the fetched data as args. And then bokeh would do the machinery around those two, to fetch data and update all sessions with it.

@havocp We implemented something like this (not without some rough spots) .  Each bokeh document requests data from the server using a cleaning function - With `cleaning functions.py`

From the analyst point of view
```
class dataframe_cleaner(object):
     def munge_some_df(df):
         do domething ....
         return munged_df
     
     Mapping = namedtuple('Mapping', ['in_df','out_df','func'])
     df_func_map = [Mapping(in_df ='df_from_some_data_store', out_df='munged_df', func=munge_some_df), ... ]
```
Our  typical bokeh directory contains:

1. main.py
2. Some .py as a release valve for analyst
3. cleaning_functions.py

And the server subscribes to data using `def initalize_data_loaders`
```
def initalize_data_loaders(srv, files_to_watch):

   #loop through the files given
    for key, path in files_to_watch.items(): 
		threading.Thread(target=_watch_a_file, args=(key, path, srv)).start()
    return
```
Are we trying to answer -  Is this functionality useful?  **Or** Does this belong as part of bokeh?
if the question is the former, some user cases have already been established 'out in the wild'.

**Comment by adam-hubbell** [comment]
@havoc, that's quite similar to what I've implemented in our system, but from outside bokeh. With some combination of you, Bryan, and Sarah's blessing, I'd be happy to work on implementing that API.

**Comment by bigreddot** [comment]
> What I would like to do is to use normal tornado callbacks within main.py so that I could, for example, use Tornado Events to trigger actions in all of the main.py instances rather than use Bokeh's PeriodicCallbacks. This would increase responsiveness and probably decrease extraneous load. This currently doesn't work because Bokeh uses locks and so doesn't play nicely with normal tornado coroutines.

> As someone familiar with Tornado

@mrocklin I would venture that being very familiar with tornado is by far the exception, not the rule. That's why we have the locks. But if there's a way to get out of the way for "super users" I'd support that. That said, have you seen this?

http://bokeh.pydata.org/en/latest/docs/user_guide/server.html#updating-from-unlocked-callbacks

It should allow you to schedule work without document locks, with the caveat that any doc updates "at the end" get rolled into a `next_tick_callback` (which is locked)

**Comment by bigreddot** [comment]
@adam-hubbell @gryBox @havocp @mrocklin is there some group of you that could participate on a Wiki outline/design page for this as a start? I can see the need for something like this but I think it's important to enumerate requirements. It would be good to see a complete example (even if it is a toy one) of how things can currently be done, along with how it could be done with the proposed API. Basically, I think this is involved enought that it warrants a "BEP"

**Comment by havocp** [comment]
Here is the email I mentioned that sketches it out a little:
https://groups.google.com/a/continuum.io/d/msg/bokeh/qVVfCUu_eow/cDZjQcVbBQAJ

A good next step might be to build a real working example, I hadn't even tested the code in that email, and bokeh has changed a bit since then too.

From there maybe someone could write up a proposal?

**Comment by gryBox** [comment]
@havocp @bigreddot take a look at this [toy example ](https://github.com/gryBox/wholesale-food-co.). `python 3.5=>`

> is there some group of you that could participate on a Wiki outline/design page for this as a start?
Yes - But be forewarned that I am still quite a novice on this stuff.

**Comment by bigreddot** [comment]
@gryBox that's a pretty involved demo for purposes of a proposal, includes unrelated parts about starting bokeh server programmatically on a custom ioloop, a fair chunk of code specific to the actual data. I was hoping for something even more pared down, possibly with synthetic data, that really shows just the concept in question in isolation.

**Comment by adam-hubbell** [comment]
@bigreddot I've got a better, simpler example in the works, I'll post it here today or tomorrow

**Comment by adam-hubbell** [comment]
[data_passing_example.zip](https://github.com/bokeh/bokeh/files/478305/data_passing_example.zip)

Here it is. It's about 180 lines total, but the code is less than 80 lines, the rest are comments. This was developed on windows 10 using python 3.5, but I don't think anything windows specific, or 3.5 specific. To run the app simply run Server.py, and to see the table change, change some values in the csv file. (Also note that it will fail with a permissions error if you have the csv open and unsaved before the server has started, so wait until it prints 'go to localhost:8790/' before messing with the csv.) 
@havocp @bigreddot

**Comment by sigurdurb** [comment]
@adam-hubbell 
I was wondering if there was any news of this being implemented into the bokeh codebase ? 
Your datapassingexample was very helpful, I added support for multiple apps and destroying inactive sessions but have yet to test it thoroughly and hopefully It should work with autoload_server since I am embedding them in a django app. If anybody knows about any more examples like this it would be very helpful. My need for this is on a streaming weather application so my df at start will be last 48 hours of some weather data. Send that as quickly as possible to each new session. Then for efficiency I am looking at streaming for each active session only the latest data points.

**Comment by bryevdv** [comment]
Milestoned to 4.0 per active dev branch but I expect to backport the related PR to the 3.10 branch for earlier release. 

I realize ten years is probably too late for the OP 😄  But this discussion was very useful to finally get a reasonable feature added, one that fell out much more more sensibly in the context of the new ASGI deployment option.
